src/main/java/com/limechain/trie/DiskTrieService.java

```diff
@@ -0,0 +1,1112 @@
+     *
+     * @param key the key path for the sought storage value.
+     * @return An {@link Optional} with the found storage value or an empty optional otherwise.
+     */
+    public Optional<byte[]> findStorageValue(Nibbles key) {
+        Optional<PendingTrieNodeChange> change = trieChanges.getFromCache(key);
+        return change.<Optional<byte[]>>map(pendingTrieNodeChange ->
+                pendingTrieNodeChange instanceof PendingInsertUpdate update
+                    ? Optional.ofNullable(update.value())
+                    : Optional.empty())
+            .orElseGet(() -> traverseTrie(trieMerkleRoot, key) instanceof TraversalResult.Found found
+                ? Optional.ofNullable(found.getFoundNode().getValue())
+                : Optional.empty());
+    }
+
+    /**
+     * This method traverses the trie from the root (if present in cache starts from there, otherwise from disk) and
+     * aims to find the closest following key in a lexicographic manner.
+     *
+     * @param key the key path whose closest lexicographic successor is needed.
+     * @return An {@link Optional} with the found key or and empty one if not found.
+     */
+    public Optional<Nibbles> getNextKey(Nibbles key) {
+        PendingInsertUpdate cachedRoot = trieChanges.getRoot().orElse(null);
+        TrieNodeData root = cachedRoot != null
+            ? new TrieNodeData(cachedRoot.value() != null,
+            cachedRoot.partialKey(),
+            cachedRoot.childrenMerkleValues(),
+            cachedRoot.value(),
+            null,
+            (byte) cachedRoot.stateVersion().asInt())
+            : trieStorage.getTrieNodeFromMerkleValue(trieMerkleRoot);
+
+        if (root == null) {
+            return Optional.empty();
+        }
+
+        return findNextKey(key, Nibbles.EMPTY, root);
+    }
+
+    private Optional<Nibbles> findNextKey(Nibbles prefix, Nibbles currentKey, TrieNodeData node) {
+        if (node.getValue() != null && currentKey.compareTo(prefix) > 0) {
+            return Optional.of(currentKey);
+        }
+
+        for (Nibble nibble : Nibbles.ALL) {
+            byte[] childMerkle = node.getChildrenMerkleValues().get(nibble.asInt());
+            if (childMerkle == null) {
+                continue;
+            }
+
+            TrieNodeData cachedChild = getCachedChildAtIndex(currentKey, nibble);
+            TrieNodeData childNode = cachedChild != null
+                ? cachedChild
+                : trieStorage.getTrieNodeFromMerkleValue(childMerkle);
+
+            Nibbles nextPath = currentKey.add(nibble).addAll(childNode.getPartialKey());
+            Optional<Nibbles> result = findNextKey(prefix, nextPath, childNode);
+            if (result.isPresent()) {
+                return result;
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    public void upsertNode(Nibbles key, byte[] storageValue, StateVersion stateVersion) {
```

Comment:
Add Javadoc to this function as well as it gets used a lot and the name is a bit ambiguous.